### PR TITLE
803 community selection page of registration process 2

### LINKED ISSUE #803: 🚀 [Feature] community selection page as 1st page of registration process
## 🚀 Feature

1. new page to select a community
2. shows a single selectable list with the names of all communities
3. the names of all communities and their registration URL has to be requested per service invocation at the community instanz ( see community creation process)
4. shows an Abbruch and Weiter Button
5. Weiter Button is enabled as soon a community is selected
6. optional: a given PID parameter check per request invocation against the GDT-Server, returns a valid PID - the given one or the default PID of the Förderkreis
7. pressing the Weiter button invokes the registration process on the selected community incuding the mandatory PID

### PR DESCRIPTION
**Pull request by ogerly**
## 🍰 Pullrequest
Registration in different communities. 

there are currently two newly created files in the app. 
1. an overview page for the current community with redirection to the current registration page.
2. an overview page to select another community. 

existing files that have changed. 
3. login.vue
4. register.vue
5. ../locales/en.json
6. ../locales/en.json

the stimulation of the federation 
7. json file : frontend/public/json-example/communities.json

### Issues
- fixes #803 
 
 
FireShot Capture 587 - Gradido Account - localhost

FireShot Capture 588 - Gradido Account - localhost

FireShot Capture 589 - Gradido Account - localhost

FireShot Capture 590 - Gradido Account - localhost

### CODE REVIEW

File: frontend/src/views/Pages/Login.vue [hunk 1]
```diff
@@ -19,8 +19,13 @@
           <b-card no-body class="border-0 mb-0" style="background-color: #ebebeba3 !important">
             <b-card-body class="p-4">
               <div class="text-center text-muted mb-4">
-                <small>{{ $t('login') }}</small>
+                <b>{{ $store.state.community.name }}</b>
```
Comment (Mogge) [inline frontend/src/views/Pages/Login.vue:22]: Bitte testen ob das da ist

File: frontend/src/views/Pages/Login.vue [hunk 2]
```diff
@@ -38,13 +43,17 @@
...
               </router-link>
             </b-col>
-            <b-col cols="6" class="text-right" v-show="allowRegister">
-              <router-link to="/register">
+            <b-col
+              cols="6"
+              class="text-center text-sm-right col-12 col-sm-6"
+              v-show="allowRegister"
+            >
+              <router-link to="/regist-community" class="mt-3">
```
Comment (Mogge) [inline frontend/src/views/Pages/Login.vue:56]: register-community

File: frontend/src/views/Pages/Register.spec.js [hunk 3]
```diff
@@ -127,6 +133,18 @@ describe('Register', () => {
...
+      it('has a link "Choose another community"', () => {
+        expect(wrapper.find('.test-button-another-community').text()).toEqual(
+          'community.choose-another-community',
+        )
+      })
+
+      it('links to /select-community when clicking "Choose another community"', () => {
+        expect(wrapper.find('.test-button-another-community').props().to).toBe('/select-community')
+      })
+    })
```
Comment (Mogge) [inline frontend/src/views/Pages/Register.spec.js:146]: Top!

File: frontend/src/views/Pages/Register.vue [hunk 4]
```diff
@@ -13,15 +13,21 @@
...
     <b-container class="mt--8 p-1">
       <!-- Table -->
+
       <b-row class="justify-content-center">
         <b-col lg="6" md="8">
           <b-card no-body class="border-0" style="background-color: #ebebeba3 !important">
             <b-card-body class="p-4">
               <div class="text-center text-muted mb-4">
-                <small>{{ $t('signup') }}</small>
+                <b>{{ $store.state.community.name }}</b>
```
Comment (Mogge) [inline frontend/src/views/Pages/Register.vue:26]: Testen

File: frontend/src/views/Pages/RegisterCommunity.vue [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    <div class="p-3">
+      <b-container>
+        <div class="text-center mb-7 header">
+          <b-row class="justify-content-center">
+            <b-col xl="5" lg="6" md="8" class="px-2">
+              <h1>{{ $store.state.community.name }}</h1>
+              <p class="text-lead">
+                {{ $store.state.community.description }}
+              </p>
+              <p class="text-lead community-location">Location: {{ $store.state.community.url }}</p>
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterCommunity.vue:13]: Location sollte übersetzt werdem

File: frontend/src/views/Pages/RegisterSelectCommunity.vue [hunk 6]
```diff
@@ -0,0 +1,79 @@
+<template>
+  <div id="register-select-community">
+    <b-container class="text-center">
+      <div class="pb-3">{{ $t('community.current-community') }}</div>
+
+      <div v-for="community in communitys.community" :key="community.name">
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.vue:6]: communities

File: frontend/src/views/Pages/RegisterSelectCommunity.vue [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        <b-card
+          v-show="community.name === $store.state.community.name"
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.vue:8]: Können wir über ein computed die Community finden, die hier angezeigt werden soll? Dann habe wir die Schleife hier nicht und können auf das if versichten

File: frontend/src/views/Pages/RegisterSelectCommunity.vue [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            {{ $t('community.continue-to-registration') }}
+          </b-button>
+        </b-card>
+      </div>
+
+      <hr />
+
+      <div>{{ $t('community.other-communities') }}</div>
+
+      <div v-for="community in communitys.community" :key="community.id" class="pb-3">
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.vue:26]: communities

File: frontend/src/views/Pages/RegisterSelectCommunity.vue [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        <b-card v-show="community.name != $store.state.community.name" bg-variant="secondary">
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.vue:27]: Warum nicht v-if? Gibt es einen Grund die aktuelle Community hier unsichtar zu rendern?

File: frontend/src/views/Pages/RegisterSelectCommunity.vue [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+          <b>{{ community.name }}</b>
+          <br />
+          {{ community.description }}
+          <br />
+          Location:
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.vue:32]: Übersetzen

File: frontend/src/views/Pages/RegisterSelectCommunity.vue [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+      </div>
+
+      <div class="text-center py-lg-4">
+        <b-button variant="outline-secondary" to="/login">{{ $t('back') }}</b-button>
+      </div>
+    </b-container>
+  </div>
+</template>
+<script>
+import MyCommunitys from '../../../public/json-example/communitys.json'
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.vue:52]: communites

File: frontend/src/views/Pages/UserProfile/UserCard.vue [hunk 12]
```diff
@@ -21,7 +21,7 @@
       </div>
       <div>
         <span class="heading">--</span>
-        <span class="description">{{ $t('community') }}</span>
+        <span class="description">{{ $t('community.community') }}</span>
```
Comment (Mogge) [inline frontend/src/views/Pages/UserProfile/UserCard.vue:24]: Testen?

File: frontend/public/json-example/communitys.json [hunk 13]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (Mogge) [inline frontend/public/json-example/communitys.json:1]: Der Name sollte communities sein

File: frontend/src/views/Pages/RegisterCommunity.spec.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+      it('has a button "Back to Login?"', () => {
+        expect(wrapper.findAll('a').at(2).text()).toEqual('back')
+      })
+      it('button links to /login when clicking "Back to Login"', () => {
+        expect(wrapper.findAll('a').at(2).props().to).toBe('/login')
+      })
+    })
+  })
+})
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterCommunity.spec.js:80]: Super!

File: frontend/src/views/Pages/RegisterSelectCommunity.vue [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+export default {
+  name: 'registerSelectCommunity',
+  data() {
+    return {
+      communitys: MyCommunitys,
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.vue:58]: communities

**Review by Mogge — CHANGES_REQUESTED** [review]

File: frontend/src/views/Pages/RegisterSelectCommunity.vue (hunk 7 above)
Comment (ogerly) [reply frontend/src/views/Pages/RegisterSelectCommunity.vue]: noch offen

File: frontend/public/json-example/communities.json [hunk 16]
```diff
@@ -0,0 +1,7 @@
+{
+  "community": [
+    {"id": 1, "name": "Gradido Entwicklung", "url": "http://localhost:3000/vue/", "description":"Die lokale Entwicklungsumgebung von Gradido.", "registerUrl": "http://localhost:3000/vue/register"},
+    {"id": 2, "name": "Gradido Staging", "url": "https://stage1.gradido.net/vue/", "description":"Der Testserver der Gradido Akademie.", "registerUrl": "https://stage1.gradido.net/vue/register"},
+    {"id": 3, "name": "Gradido-Akademie", "url": "https://gdd1.gradido.com/vue/", "description":"Freies Institut für Wirtschaftsbionik.", "registerUrl": "https://gdd1.gradido.com/vue/register"}
+  ]
+}
```
Comment (Elweyn) [inline frontend/public/json-example/communities.json:7]: This should be ported to the apollo server! So we can test it in the frontend.

File: frontend/src/store/store.js [hunk 17]
```diff
@@ -68,6 +68,12 @@ export const store = new Vuex.Store({
     description: '',
     token: null,
     newsletterState: null,
+    community: {
+      name: 'Gradido Entwicklung',
+      url: 'http://localhost:3000/vue/',
+      registerUrl: 'http://localhost:3000/vue/register',
+      description: 'Die lokale Entwicklungsumgebung von Gradido.',
+    },
```
Comment (Elweyn) [inline frontend/src/store/store.js:76]: Would be nice to get this out of the .env file

**Review by Elweyn — CHANGES_REQUESTED** [review]

File: frontend/public/json-example/communities.json (hunk 16 above)
Comment (ogerly) [reply frontend/public/json-example/communities.json:7]: yes pleace

File: frontend/src/store/store.js (hunk 17 above)
Comment (ogerly) [reply frontend/src/store/store.js:76]: i need a little help ...

Comment (ulfgebhardt) [reply frontend/src/store/store.js:76]: Server's `.env` file

File: backend/src/config/index.ts [hunk 18]
```diff
@@ -30,9 +30,18 @@ const klicktipp = {
   KLICKTIPP_APIKEY_EN: process.env.KLICKTIPP_APIKEY_EN || 'SomeFakeKeyEN',
 }
 
+const community = {
+  COMMUNITY_NAME: process.env.COMMUNITY_NAME || 'Gradido Entwicklung',
+  COMMUNITY_URL: process.env.COMMUNITY_URL || 'https://gradido.net/',
```
Comment (Mogge) [inline backend/src/config/index.ts:35]: `http://localhost/vue` ?

File: backend/src/config/index.ts [hunk 19]
```diff
@@ -30,9 +30,18 @@ const klicktipp = {
...
+  COMMUNITY_REGISTER_URL:
+    process.env.COMMUNITY_REGISTER_URL || 'http://localhost:3000/vue/register',
```
Comment (Mogge) [inline backend/src/config/index.ts:37]: remove the port (`:3000`)

File: backend/src/graphql/model/Community.ts [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+  @Field(() => Number)
+  id: number
+
+  @Field(() => String)
+  name: string
+
+  @Field(() => String)
+  url: string
+
+  @Field(() => String)
```
Comment (Mogge) [inline backend/src/graphql/model/Community.ts:26]: nullable?

File: backend/src/graphql/resolver/CommunityResolver.ts [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+  @Query(() => [Community])
+  async communities(): Promise<Community[]> {
+    const communities: Community[] = []
+
+    communities.push(
+      new Community({
+        id: 1,
+        name: 'Gradido Entwicklung',
+        description: 'Die lokale Entwicklungsumgebung von Gradido.',
+        url: 'http://localhost:3000/vue/',
```
Comment (Mogge) [inline backend/src/graphql/resolver/CommunityResolver.ts:29]: remove the port

File: backend/src/graphql/resolver/CommunityResolver.ts [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+        registerUrl: 'http://localhost:3000/vue/register-community',
```
Comment (Mogge) [inline backend/src/graphql/resolver/CommunityResolver.ts:30]: no port

File: backend/src/graphql/resolver/CommunityResolver.ts [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+      }),
+      new Community({
+        id: 2,
+        name: 'Gradido Staging',
+        description: 'Der Testserver der Gradido Akademie.',
```
Comment (Mogge) [inline backend/src/graphql/resolver/CommunityResolver.ts:35]: hyphen? Gradido-Akademie

File: frontend/src/views/Pages/Login.spec.js [hunk 24]
```diff
@@ -30,6 +36,15 @@ describe('Login', () => {
     $t: jest.fn((t) => t),
     $store: {
       dispatch: mockStoreDispach,
+      commit: mockStoreCommit,
+      state: {
+        community: {
+          name: 'Gradido Entwicklung',
+          url: 'http://localhost:3000/vue/',
```
Comment (Mogge) [inline frontend/src/views/Pages/Login.spec.js:43]: no port, just for detail

File: frontend/src/views/Pages/Login.spec.js [hunk 25]
```diff
@@ -62,12 +77,33 @@ describe('Login', () => {
       expect(wrapper.find('div.login-form').exists()).toBeTruthy()
     })
 
+    it('calls the communityInfo', () => {
```
Comment (Mogge) [inline frontend/src/views/Pages/Login.spec.js:80]: `commits the community info to the store`. I miss a test that the API was called before.

File: frontend/src/views/Pages/Login.spec.js [hunk 26]
```diff
@@ -62,12 +77,33 @@ describe('Login', () => {
...
+    })
+
     describe('Login header', () => {
       it('has a welcome message', () => {
         expect(wrapper.find('div.header').text()).toBe('Gradido site.login.community')
       })
     })
 
+    describe('Community Data', () => {
+      it('has a Community name?', () => {
```
Comment (Mogge) [inline frontend/src/views/Pages/Login.spec.js:96]: why a question mark?

File: frontend/src/views/Pages/Login.spec.js [hunk 27]
```diff
@@ -62,12 +77,33 @@ describe('Login', () => {
...
+        expect(wrapper.find('.test-communitydata b').text()).toBe('Gradido Entwicklung')
+      })
+
+      it('has a Community description?', () => {
```
Comment (Mogge) [inline frontend/src/views/Pages/Login.spec.js:100]: ?

File: frontend/src/views/Pages/Login.spec.js [hunk 28]
```diff
@@ -135,10 +173,15 @@ describe('Login', () => {
           await flushPromises()
           await wrapper.find('form').trigger('submit')
           await flushPromises()
+          apolloQueryMock.mockResolvedValue({
+            data: {
+              login: 'token',
```
Comment (Mogge) [inline frontend/src/views/Pages/Login.spec.js:178]: This was wrong before. The login now returns a user object.

File: frontend/src/views/Pages/Login.vue [hunk 29]
```diff
@@ -103,6 +112,21 @@ export default {
...
+    async onCreated() {
+      this.$apollo
+        .query({
+          query: communityInfo,
+        })
+        .then((result) => {
+          this.$store.commit('community', result.data.getCommunityInfo)
+        })
+        .catch((error) => {
+          this.$toasted.error(error)
```
Comment (Mogge) [inline frontend/src/views/Pages/Login.vue:124]: Test is missing

File: frontend/src/views/Pages/Register.spec.js [hunk 30]
```diff
@@ -26,6 +26,12 @@ describe('Register', () => {
       state: {
         email: 'peter@lustig.de',
         language: 'en',
+        community: {
+          name: 'Gradido Entwicklung',
+          url: 'http://localhost:3000/vue/',
```
Comment (Mogge) [inline frontend/src/views/Pages/Register.spec.js:31]: no port

File: frontend/src/views/Pages/RegisterCommunity.spec.js [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+  const mocks = {
+    $i18n: {
+      locale: 'en',
+    },
+    $t: jest.fn((t) => t),
+    $store: {
+      state: {
+        community: {
+          name: 'Gradido Entwicklung',
+          url: 'http://localhost:3000/vue/',
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterCommunity.spec.js:18]: no port

File: frontend/src/views/Pages/RegisterSelectCommunity.spec.js [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+  const mocks = {
+    $i18n: {
+      locale: 'en',
+    },
+    $t: jest.fn((t) => t),
+    $store: {
+      state: {
+        community: {
+          name: 'Gradido Entwicklung',
+          url: 'http://localhost:3000/vue/',
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.spec.js:19]: no port

File: frontend/src/views/Pages/RegisterSelectCommunity.spec.js [hunk 33]
```diff
@@ -0,0 +1,40 @@
+import { mount } from '@vue/test-utils'
+import RegisterSelectCommunity from './RegisterSelectCommunity'
+// const testObject = require('../../../public/json-example/communities.json')
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.spec.js:3]: please remove

File: frontend/src/views/Pages/RegisterSelectCommunity.vue [hunk 34]
```diff
@@ -0,0 +1,105 @@
...
+      communities: [],
+      pending: true,
+    }
+  },
+  methods: {
+    async getCommunities() {
+      const loader = this.$loading.show({
+        container: this.$refs.header,
+      })
+      this.$apollo
```
Comment (Mogge) [inline frontend/src/views/Pages/RegisterSelectCommunity.vue:69]: Test is missing

**Review by Mogge — CHANGES_REQUESTED** [review]
Ein paar Kleinigkeiten. Gute Arbeit!

File: frontend/src/views/Pages/Login.spec.js (hunk 25 above)
Comment (ulfgebhardt) [reply frontend/src/views/Pages/Login.spec.js:80]:
```suggestion
    it('commits the community info to the store', () => {
```

**Review by Elweyn — APPROVED** [review]

**Pull request merged**